### PR TITLE
Add GePaAlignmentOptimizer for judge instruction optimization

### PR DESCRIPTION
**Pull request by alkispoly-db**
<details><summary>🛠 DevTools 🛠</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/alkispoly-db/mlflow/pull/19882?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19882/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19882/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/19882/merge
```

</p>
</details>

### Related Issues/PRs

N/A

### What changes are proposed in this pull request?

This PR implements `GEPAAlignmentOptimizer`, a new alignment optimizer for MLflow judges that uses the GEPA (Genetic-Pareto) algorithm to optimize judge instructions by learning from human feedback in traces.

**Key Features:**
- Extends `DSPyAlignmentOptimizer` base class, following the same pattern as `SIMBAAlignmentOptimizer`
- Auto-calculates optimization budget (4x training examples by default)
- Adds `feedback_value_type` as abstract property on `Judge` base class

**Implementation:**
- Main class: `GEPAAlignmentOptimizer` extending `DSPyAlignmentOptimizer` (~140 lines)
- Utility functions in `dspy_utils.py` for demo formatting, input field handling
- Follows established patterns from `SIMBAAlignmentOptimizer`

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

**Test Coverage:**
- 121 comprehensive unit tests across optimizer modules
- Tests cover: import errors, full workflows, custom parameters, edge cases
- All tests passing with full ruff and clint compliance

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

**Release Note:**
Adds `GEPAAlignmentOptimizer` for optimizing judge instructions using the GEPA algorithm. This optimizer learns from human feedback in traces to iteratively improve judge performance through genetic-pareto optimization. Users can now align judges by calling `optimizer.align(judge, traces)` where traces contain human assessments.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [x] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/genai/judges/optimizers/gepa.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    raise MlflowException(
+        "DSPy library is required but not installed. Please install it with: pip install dspy",
+        error_code=INTERNAL_ERROR,
+    )
+
+_logger = logging.getLogger(__name__)
+
+
+@experimental(version="3.8.0")
+class GePaAlignmentOptimizer(DSPyAlignmentOptimizer):
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/gepa.py]: GePaAlignmentOptimizer -> GEPAAlignmentOptimizer

File: mlflow/genai/judges/optimizers/gepa.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+)
+from mlflow.protos.databricks_pb2 import INTERNAL_ERROR
+from mlflow.utils.annotations import experimental
+
+# Import dspy - raise exception if not installed
+try:
+    import dspy
+except ImportError:
+    raise MlflowException(
+        "DSPy library is required but not installed. Please install it with: pip install dspy",
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/gepa.py]:
```suggestion
        "DSPy library is required but not installed. Please install it with: `pip install dspy`",
```

File: mlflow/genai/judges/optimizers/gepa.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        optimizer = dspy.GEPA(**optimizer_kwargs)
+
+        self._logger.info(
+            f"Starting GEPA optimization with {len(examples)} examples "
+            f"and max {self._max_metric_calls} metric calls"
+        )
+
+        with suppress_verbose_logging("dspy.teleprompt.gepa.gepa"):
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/gepa.py]: I personally think the logging of GEPA is actually helpful. Without this users won't know the progress, correct? If so, users might feel nervous to wait for ~30 minutes without progress information.

File: mlflow/genai/judges/optimizers/gepa.py (hunk 2 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/gepa.py]: Done.

File: mlflow/genai/judges/optimizers/gepa.py (hunk 3 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/gepa.py]: We suppress verbose output from other optimizers, so I think this is consistent. Let's tackle this in a follow-up PR to add a flag for verbose output to the optimizers.

File: mlflow/genai/judges/optimizers/gepa.py (hunk 1 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/gepa.py]: Done.

File: mlflow/genai/judges/optimizers/dspy.py [hunk 4]
```diff
@@ -71,66 +74,104 @@ def __init__(self, model: str | None = None, **kwargs):
...
+            program: The DSPy Predict program to optimize
             examples: Examples for optimization (implementations decide how to split)
             metric_fn: Metric function for optimization
 
         Returns:
-            Optimized DSPy program
+            Optimized DSPy Predict program
+        """
+
+    def _create_judge_from_optimized_program(
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/dspy.py]: nit: function wise, this method converts a dspy program to a judge regardless of the optimization status, so can we rename it as `_create_judge_from_dspy_program`?

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 5]
```diff
@@ -32,6 +37,66 @@
...
+    try:
+        if _logger.getEffectiveLevel() > threshold_level:
+            logger.setLevel(logging.WARNING)
+        yield
+    finally:
+        logger.setLevel(original_level)
+
+
+def create_gepa_metric_adapter(
+    metric_fn: Any,
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: can we update the type hint to Callable?

File: mlflow/genai/judges/optimizers/dspy_utils.py [hunk 6]
```diff
@@ -56,9 +121,56 @@ def construct_dspy_lm(model: str):
         return AgentEvalLM()
     else:
         model_litellm = convert_mlflow_uri_to_litellm(model)
+        if api_base := _get_api_base_for_model(model):
+            return dspy.LM(model=model_litellm, api_base=api_base)
         return dspy.LM(model=model_litellm)
 
 
+def _get_api_base_for_model(model: str) -> str | None:
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/dspy_utils.py]: _get_api_base_for_model -> _get_databricks_api_base

File: mlflow/genai/scorers/builtin_scorers.py [hunk 7]
```diff
@@ -326,6 +326,11 @@ def instructions(self) -> str:
         Get the instructions of what this scorer evaluates.
         """
 
+    @property
+    def feedback_value_type(self) -> Any:
+        """Get the type of the feedback value."""
+        return str
```
Comment (TomeHirata) [inline mlflow/genai/scorers/builtin_scorers.py]: Could we specify `Listeral["yes", "no", "unknown"]` to be more accurate? Or does it cause any issues?

File: mlflow/genai/judges/optimizers/gepa.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            gepa_kwargs: Additional keyword arguments to pass directly to dspy.GEPA().
+            **kwargs: Additional keyword arguments passed to parent class
+        """
+        super().__init__(model=model, **kwargs)
+        self._max_metric_calls = max_metric_calls
+        self._gepa_kwargs = gepa_kwargs or {}
+
+    def _dspy_optimize(
+        self,
+        program: "dspy.Module",
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/gepa.py]:
```suggestion
        program: "dspy.Predict",
```

File: mlflow/genai/judges/optimizers/gepa.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        examples: Collection["dspy.Example"],
+        metric_fn: Callable[["dspy.Example", Any, Any | None], bool],
+    ) -> "dspy.Module":
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/gepa.py]:
```suggestion
    ) -> "dspy.Predict":
```

File: mlflow/genai/judges/optimizers/gepa.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                f"max_metric_calls not specified, using {self._DEFAULT_BUDGET_MULTIPLIER}x "
+                f"number of examples: {max_metric_calls}"
+            )
+
+        optimizer_kwargs = {
+            "metric": gepa_metric_adapter,
+            "max_metric_calls": max_metric_calls,
+            "reflection_lm": reflection_lm,
+            **self._gepa_kwargs,  # Pass through any additional GEPA parameters
+        }
```
Comment (TomeHirata) [inline mlflow/genai/judges/optimizers/gepa.py]:
```suggestion
        optimizer_kwargs = self._gepa_kwargs | {
            "metric": gepa_metric_adapter,
            "max_metric_calls": max_metric_calls,
            "reflection_lm": reflection_lm,
            **self._gepa_kwargs,  # Pass through any additional GEPA parameters
        }
```
So that metric or reflection_lm is not overwritten by `_gepa_kwargs`

**Review by TomeHirata — APPROVED** [review]
Overall LGTM, can we update the documentation?

File: mlflow/genai/judges/optimizers/dspy.py (hunk 4 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/dspy.py]: Done.

File: mlflow/genai/judges/optimizers/dspy_utils.py (hunk 5 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/dspy_utils.py]: Done.

File: mlflow/genai/judges/optimizers/gepa.py (hunk 8 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/gepa.py]: Done.

File: mlflow/genai/judges/optimizers/gepa.py (hunk 9 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/gepa.py]: Done.

File: mlflow/genai/judges/optimizers/gepa.py (hunk 10 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/gepa.py]: Good catch. Done.

File: mlflow/genai/scorers/builtin_scorers.py (hunk 7 above)
Comment (alkispoly-db) [reply mlflow/genai/scorers/builtin_scorers.py]: Built-in scorers have different conventions so "str" is the safer option. This also buys us robustness for future changes to built-in scorers (we make fewer assumptions).

File: mlflow/genai/judges/optimizers/dspy_utils.py (hunk 6 above)
Comment (alkispoly-db) [reply mlflow/genai/judges/optimizers/dspy_utils.py]: I think the method name is more generic so I 'd opt to keep it as is. I refactored the Databricks code into a specific method so in the future we can easily add similar support for other providers.
